src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensTests.cs

```diff
@@ -0,0 +1,144 @@
+                       0,     2,     1,    SemanticTokensCache.TokenTypesToIndex[LSP.SemanticTokenTypes.Operator], 0, // '}'
+                },
+                ResultId = "2"
+            };
+
+            Assert.Equal(expectedWholeDocResults.Data, wholeDocResults.Data);
+            Assert.Equal(expectedWholeDocResults.ResultId, wholeDocResults.ResultId);
+
+            // 3. Edits handler - insert newline at beginning of file
+            // Markup becomes:
+            //
+            //   // Comment
+            //   static class C { }";
+            var currentDocText = await workspace.CurrentSolution.Projects.First().Documents.First().GetTextAsync();
+            var changedSourceText = currentDocText.WithChanges(new TextChange(new TextSpan(0, 0), "\n"));
+            var docId = ((TestWorkspace)workspace).Documents.First().Id;
+            ((TestWorkspace)workspace).ChangeDocument(docId, changedSourceText);
+            UpdateSolutionProvider((TestWorkspace)workspace, workspace.CurrentSolution);
+
+            var editResults = await RunGetSemanticTokensEditsAsync(workspace.CurrentSolution, caretLocation, previousResultId: "2");
+
+            var expectedEdit = SemanticTokensEditsHandler.GenerateEdit(0, 1, new int[] { 1 });
+
+            Assert.Equal(expectedEdit, ((LSP.SemanticTokensEdits)editResults).Edits.First());
+            Assert.Equal("3", ((LSP.SemanticTokensEdits)editResults).ResultId);
+
+            // 4. Re-request whole document handler (may happen if LSP runs into an error)
+            var wholeDocResults2 = await RunGetSemanticTokensAsync(workspace.CurrentSolution, caretLocation);
+            var expectedWholeDocResults2 = new LSP.SemanticTokens
+            {
+                Data = new int[]
+                {
+                    // Line | Char | Len | Token type                                                            | Modifier
+                       1,     0,     10,   SemanticTokensCache.TokenTypesToIndex[LSP.SemanticTokenTypes.Comment],  0, // '// Comment'
+                       1,     0,     6,    SemanticTokensCache.TokenTypesToIndex[LSP.SemanticTokenTypes.Keyword],  0, // 'static'
+                       0,     7,     5,    SemanticTokensCache.TokenTypesToIndex[LSP.SemanticTokenTypes.Keyword],  0, // 'class'
+                       0,     6,     1,    SemanticTokensCache.TokenTypesToIndex[LSP.SemanticTokenTypes.Class],    (int)TokenModifiers.Static, // 'C'
+                       0,     2,     1,    SemanticTokensCache.TokenTypesToIndex[LSP.SemanticTokenTypes.Operator], 0, // '{'
+                       0,     2,     1,    SemanticTokensCache.TokenTypesToIndex[LSP.SemanticTokenTypes.Operator], 0, // '}'
+                },
+                ResultId = "4"
+            };
+
+            Assert.Equal(expectedWholeDocResults2.Data, wholeDocResults2.Data);
+            Assert.Equal(expectedWholeDocResults2.ResultId, wholeDocResults2.ResultId);
+        }
```

Comment:
i may have missed it.  but i would love some tests that don't actually ever check the raw values.  instead tehy do this:

// pseudo code:
`ApplyEdits(FullClassification(oldCode), SomeUserEdits) equals FullClassification(oldCode + someUserEdits)`

Reply:
Yep! I added three of these while addressing the last iteration of your feedback, there is this [test](https://github.com/dotnet/roslyn/pull/46407/files#diff-18039c1c8975d6f3dff5f170f34daedbR182) and the two tests below it that verify the above scenario.